FEATURE: Mixed text direction support

scossar opened the pull request:
Adds a mixed_text_direction_support setting. When enabled it:

- adds a toggle-direction button to the composer
- sets the text direction on all cooked posts
- sets the text direction on the fancy_title
- sets the text direction on category descriptions
- sets the text direction on category names
- sets the text direction on topic excerpts
- adds a dir attribute to text fields, the dir is set based on the text field's content

For titles and category descriptions it's setting the dir in an inner span. Setting it on the outer element, for example a `h1` tag, will mess up the page formatting.

There are no tests yet, but these changes shouldn't make any existing tests fail.

SamSaffron commented on app/assets/javascripts/discourse/components/d-editor.js.es6:
```diff
@@ -107,6 +108,17 @@ class Toolbar {
+        id: 'toggle-direction',
+        group: 'extras',
+        icon: 'arrows-h',
+        shortcut: 'Shift+6',
```
hmmm is this not a bit too common?

SamSaffron commented:
@eviltrout can you review? 

@scossar noticing there are no tests here can you add some?

eviltrout commented on app/assets/javascripts/discourse/components/text-field.js.es6:
```diff
@@ -1,7 +1,35 @@
 import computed from "ember-addons/ember-computed-decorators";
+import { siteDir } from "discourse/lib/text-direction";
```
You can import all three on one line:

`import { siteDir, isRTL, isLTR } from ...`

eviltrout commented on app/assets/javascripts/discourse/components/text-field.js.es6:
```diff
@@ -1,7 +1,35 @@
+
+  @computed
+  dir() {
+    if (Discourse.SiteSettings.support_mixed_text_direction) {
```
In a component you should use `this.siteSettings` instead of `Discourse.SiteSettings`

eviltrout commented on app/assets/javascripts/discourse/components/text-field.js.es6:
```diff
@@ -1,7 +1,35 @@
+  },
+
+  keyUp() {
+    if (Discourse.SiteSettings.support_mixed_text_direction) {
```
Again should be `this.siteSettings`

eviltrout commented on app/assets/javascripts/discourse/lib/text-direction.js.es6:
```diff
@@ -0,0 +1,28 @@
+const rtlChars = '\u0591-\u07FF\uFB1D-\uFDFD\uFE70-\uFEFC';
+
+export function isRTL(text) {
+  const rtlDirCheck = new RegExp('^[^'+ltrChars+']*['+rtlChars+']');
```
Rather than parsing this every time it's called, perhaps you should move `rtlDirCheck` outside this function as a constant? Then it would only be parsed once.  (Same with `ltrDirCheck` below)

The only gotcha is you have to reset `lastIndex` before you use it., but it should be more efficient.

eviltrout commented on app/assets/javascripts/discourse/lib/text-direction.js.es6:
```diff
@@ -0,0 +1,28 @@
+}
+
+export function siteDir() {
+  return $('html').hasClass('rtl') ? 'rtl' : 'ltr';
```
You should cache this result. If it's in the html of the document I don't expect it to change. You can use a variable outside the function like `let _siteDir;` then `if (!siteDir) { ... calculate it }`

eviltrout commented on app/assets/javascripts/discourse/models/topic.js.es6:
```diff
@@ -58,7 +59,13 @@ const Topic = RestModel.extend({
   @computed('fancy_title')
   fancyTitle(title) {
-    return censor(emojiUnescape(title || ""), Discourse.Site.currentProp('censored_words'));
+    let fancyTitle = censor(emojiUnescape(title || ""), Discourse.Site.currentProp('censored_words'));
```
It wasn't in here before, but can you try `this.site` instead of `Discourse.Site` here? If it works it would be a win.

scossar replied:
It doesn't work. this.site isn't defined for the categories route.

eviltrout replied:
Ugh, I should fix that at some point.

eviltrout commented on app/assets/javascripts/discourse/models/topic.js.es6:
```diff
@@ -58,7 +59,13 @@ const Topic = RestModel.extend({
-    return censor(emojiUnescape(title || ""), Discourse.Site.currentProp('censored_words'));
+    let fancyTitle = censor(emojiUnescape(title || ""), Discourse.Site.currentProp('censored_words'));
+
+    if (Discourse.SiteSettings.support_mixed_text_direction) {
```
This should be `this.siteSettings` - although the `Discourse.Site` above makes me wonder if it's not wired up properly. If it doesn't work falling back to `Discourse.SiteSettings` is fine.

scossar commented:
I've added a some tests. I'd like to improve the tests for `lib/text-direction`. My IDE goes a little nuts when I try to add mixed ltr/rtl text to a string. I'd also like to add some tests for making sure that the cooked post content is decorated correctly.

I've incorporated most of the changes that were suggested. I don't think the regular expression's `lastIndex` needs to be reset, since it's not using the global flag.

It's hard to find a good icon for the direction-toggle button, I've changed it to `exchange`. That seems a little better.

eviltrout commented:
Looks good now!

SamSaffron commented:
@eviltrout feel to merge in we really want this deployed to macudacity
